allow user to specify custom version id for flex deployments

etanshaul opened the pull request:
fixes #572 

Version override feature is implemented:

![image](https://cloud.githubusercontent.com/assets/1735744/14258586/fdd3c67e-fa70-11e5-9db3-67d01fc0a992.png)

etanshaul commented:
Just realized, I need to implement the placeholder text feature (something like 'auto' when there is no version override).

patflynn commented:
yeah I would put something like 'automatically generated', since we have
lots of horizontal space to work with.

On Mon, Apr 4, 2016 at 2:26 PM, Etan Shaul <notifications@github.com> wrote:

> Just realized, I need to implement the placeholder text feature (something
> like 'auto' when there is no version override).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/585#issuecomment-205435812>
>

chanseokoh commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java:
```diff
@@ -187,6 +199,7 @@ protected void applyEditorTo(AppEngineDeploymentConfiguration configuration)
     configuration.setAppYamlPath(appYamlPathField.getText());
     configuration.setConfigType(getConfigType());
+    configuration.setVersion(versionOverrideCheckBox.isSelected() ? versionIdField.getText() : null);
 
```
Should there be validation to exclude, e.g., white spaces, special characters?

etanshaul replied:
yes, I'll add a regex. The pattern is printed out from gcloud when an invalid version string is sent.

etanshaul commented:
Updated with placeholder text (I was working on creating a reusable JTextField subclass that adds placeholder functionality. Putting a "hardcoded" version out for review in the meantime):

![image](https://cloud.githubusercontent.com/assets/1735744/14260574/47911402-fa7a-11e5-829e-f74d110c8b12.png)

Regarding @chanseokoh's point about validation: wouldn't it be best to just let it pass through to gcloud and let it do the validation? This minimizes client logic that we need to maintain (if the validation pattern is ever changed for instance).

chanseokoh commented:
About validation: that's a nice idea. That said, I just notice that --version is wrapped in an unusual way in the actual command:

--quiet "--version=some name" --format=json app.yaml

which results in bizarre error message. It will be better to be --version="some name". I guess it's doing automatic escaping then?

etanshaul commented:
Updated placeholder logic to use a barebones JTextField subclass.

etanshaul commented:
@chanseokoh what error are you seeing? The placement of the version id hasn't changed. Could it be that you were using an invalid version pattern? I don't think whitespace is allowed.

chanseokoh commented:
I just tried an unacceptable name with a space for the sake of testing. The message is

Fetching storage object: gs://staging.springboot-maven-project.appspot.com/us.gcr.io/springboot-maven-project/default.some name:latest#1459799851639000
CommandException: No URLs matched: gs://staging.springboot-maven-project.appspot.com/us.gcr.io/springboot-maven-project/default.some
ERROR
ERROR: fetching gcs source: running gsutil: exit status 1
ERROR: (gcloud.preview.app.deploy) Error Response: [2] Build failed; check build logs for details

etanshaul commented:
Ah. I guess its GeneralCommandLine that's doing that escaping. If you try with another invalid version id, e.g. MyId (caps aren't allowed), you'll get a better error message. I'll think about a solution - ideally we wouldn't duplicate the validation that gcloud already does.

chanseokoh commented:
I think we are fine with white space. I tested with gcloud with --version="some name" and it returned the same weird message. We don't want to duplicate validation.

etanshaul commented:
Updated:
- fixed bug discussed offline with @chanseokoh where serialized state was not being restored
- added version validation: if override is selected but version is empty, it warns the user (and prevents deploy)

@chanseokoh @patflynn ptal

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeploymentRunConfigurationEditor.java:
```diff
@@ -168,15 +177,22 @@ public File get() {
           }
         }, dockerFilePathField, userSpecifiedArtifactFileSelector));
+    versionOverrideCheckBox.addItemListener(
+        new CustomFieldOverrideListener( versionOverrideCheckBox, versionIdField));
```
there's an errant white space before versionOverrideCheckBox

patflynn commented:
LGTM

chanseokoh commented:
The bug is gone. LGTM.
